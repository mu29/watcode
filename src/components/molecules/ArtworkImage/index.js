import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Text, Loading } from 'components'
import { palette } from 'services/style'

const textColor = ({ isBookmarked, theme }) =>
  palette(isBookmarked ? 'yellow.default' : 'gray.30')({ theme })

const Wrapper = styled.article`
  flex-shrink: 0;
  position: relative;
  width: 4.5rem;
  height: 6rem;
  margin-right: 0.5rem;
  overflow: hidden;
  border: 0.0625rem solid ${palette('gray.30')};
  cursor: pointer;
`

const Image = styled.img`
  position: absolute;
  width: 100%;
  height: 100%;
  filter: blur(0.1rem);
`

const CodeArea = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 0.375rem;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${palette('gray.100')};
`

const StyledText = styled(Text)`
  color: ${textColor};
  transition: color 0.2s;

  ${Wrapper}:hover & {
    color: ${palette('yellow.default')};
  }
`

class ArtworkImage extends Component {
  toggleBookmark = (event) => {
    const {
      email,
      isBookmarked,
      createBookmark,
      deleteBookmark,
    } = this.props
    event.preventDefault()

    if (isBookmarked) {
      deleteBookmark()
    } else {
      createBookmark(email)
    }
  }

  render() {
    const {
      id,
      url,
      isLoading,
      isBookmarked,
    } = this.props
    return (
      <Wrapper { ...this.props } onClick={ this.toggleBookmark }>
        <Image src={ `https://${url}` } />
        <CodeArea>
          <Loading isLoading={ isLoading } size={ 12 } color="white.default" compact>
            <StyledText isBookmarked={ isBookmarked } fontSize={ 12 } small>#{ id }</StyledText>
          </Loading>
        </CodeArea>
      </Wrapper>
    )
  }
}

ArtworkImage.propTypes = {
  id: PropTypes.number.isRequired,
  url: PropTypes.string.isRequired,
  email: PropTypes.string,
  isLoading: PropTypes.bool.isRequired,
  isBookmarked: PropTypes.bool.isRequired,
  createBookmark: PropTypes.func.isRequired,
  deleteBookmark: PropTypes.func.isRequired,
}

ArtworkImage.defaultProps = {
  email: undefined,
}

export default ArtworkImage
