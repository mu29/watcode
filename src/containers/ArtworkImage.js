import React from 'react'
import { connect } from 'react-redux'
import { ArtworkImage } from 'components'
import { createBookmarkActions, deleteBookmarkActions } from 'store/actions'
import { getIsLoading } from 'store/selectors'

const ArtworkImageContainer = props => <ArtworkImage { ...props } />

const mapStateToProps = (state, { id }) => ({
  email: state.auth.user.email,
  isBookmarked: state.bookmark.bookmarks.find(b => b === id) !== undefined,
  isLoading: getIsLoading(state, createBookmarkActions.type)
    || getIsLoading(state, deleteBookmarkActions.type),
})

const mapDispatchToProps = (dispatch, { id, email, isBookmarked }) => ({
  toggleBookmark: () => dispatch(isBookmarked ?
    deleteBookmarkActions.request({ id }) :
    createBookmarkActions.request({ id, email })),
})

export default connect(mapStateToProps, mapDispatchToProps)(ArtworkImageContainer)
