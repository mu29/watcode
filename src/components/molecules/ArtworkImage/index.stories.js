import React from 'react'
import { storiesOf } from '@storybook/react'
import ArtworkImage from '.'

storiesOf('Molecules/ArtworkImage', module)
  .add('기본', () => (
    <ArtworkImage id={ 112047 } url="https://laftelimage.blob.core.windows.net/items/thumbs/large/3390a970-59de-4d04-a705-6814d2f2446a.jpg" />
  ))
