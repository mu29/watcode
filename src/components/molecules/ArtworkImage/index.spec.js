import React from 'react'
import ArtworkImage from '.'

const attributres = {
  code: 112047,
  url: 'https://laftelimage.blob.core.windows.net/items/thumbs/large/395db1d0-ead8-407f-a316-04aee1e668a7.jpg',
}

const component = (props = {}) => withTheme(<ArtworkImage { ...attributres } { ...props } />)

describe('<ArtworkImage />', () => {
  it('정상적으로 렌더링된다', () => {
    const wrapper = render(component())
    expect(wrapper).toMatchSnapshot()
  })
})
