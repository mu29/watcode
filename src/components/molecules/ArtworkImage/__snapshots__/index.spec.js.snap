// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArtworkImage /> 정상적으로 렌더링된다 1`] = `
.c0 {
  position: absolute;
  width: 100%;
  height: 100%;
  -webkit-filter: blur(0.1rem);
  filter: blur(0.1rem);
}

.c1 {
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<span
  componentname="Link"
>
  <img
    class="c0"
    src="https://laftelimage.blob.core.windows.net/items/thumbs/large/395db1d0-ead8-407f-a316-04aee1e668a7.jpg"
  />
  <div
    class="c1"
  >
    <span
      componentname="Text"
    >
      #
    </span>
    <span
      componentname="Text"
    >
      112047
    </span>
  </div>
</span>
`;
